Cxx fixes

thehajime opened the pull request.

retrage committed:
add g++ rumprun wrapper

Signed-off-by: Akira Moroo <retrage01@gmail.com>

retrage committed:
build libcxx even if GNU toolchains

Signed-off-by: Akira Moroo <retrage01@gmail.com>

retrage committed:
travis: use language cpp

Signed-off-by: Akira Moroo <retrage01@gmail.com>

retrage committed:
travis: fix cxx arm32 build

The old g++ (5.4) can not build libcxx. To avoid this issue, this commit
changes to update distribution to bionic and use cross toolchains from
apt. The installed gcc (7.4.0) generates code for ARMv7 by default. This
causes __ARM_ARCH issue and error in link-time in musl sigsetjmp.

Signed-off-by: Akira Moroo <retrage01@gmail.com>

thehajime commented:
To:@retrage
lkl へのパッチとかあるので、そのまま merge はしませんが、質問等させてください。commit の整形等は僕の方でおこないます。

- arm関連の修正は、cppサポートとは独立でしょうか?
- 今回対応って、g++で llvm/libcxx等をビルドするって感じ?
- pie disable は、どのような事象への対応でしょうか?
- pie disable のコミットメッセージの中の「This commit is identical with commit
f02d67a.」ってどういう意味でしょうか?

retrage commented:
このbranchの主な目的はgcc/g++でのlibcxxのビルドを有効にすることです．

1. d127cbf3 と 4f9de222 は主にarm32でのC++対応のためのコミットです．

2. e004ba80 と 5f5e5b16 はg++でもlibcxxをビルドするようにしたコミットです．
(手元環境ではうまくいったのですが，古い環境だとうまくいかなかった覚えがあります．なのでbionicにしています)

3. 比較的新しいgccだと`--enable-default-pie`がコンパイラのビルド時のconfigurationで有効になっており，そのコンパイラでビルドするとオプションなしでも`-pie`をつけたのと同じようにコンパイルがされます．このとき`-Wl,-r`をつけてビルドした場合に`-pie`と`-Wl,-r`を同時に指定したことになり，リンクエラーになります．これは`build.sh`の`rumpkernel.o`を生成するリンク時に発生します．"This commit is identical with commit f02d67a." は以前のコミットと等価である，ぐらいの意味のつもりでした．
`-no-pie`については結論としてはsolo5 branchでの修正が良さそうです．

このPRでは1.と2.のコミットを取り込むのがいいと思います．

3. についてsolo5 branchでの修正を取り込んだコミットをcxx-fixesに入れた方がいいでしょうか？

retrage commented:
solo5での修正を取り込んだコミットを入れました

thehajime commented:
Memo: https://twitter.com/gnutools/status/1245362276460191749?s=21

retrage committed:
cxx: fix compiler options

Some old g++ requires -std=c++11 option explicitly.
Adding EXTRA_LDSCRIPT_CC to COMPILER_FLAGS causes invalid ELF
generation.

Signed-off-by: Akira Moroo <retrage01@gmail.com>

retrage committed:
add__dso_handle definition for g++

Signed-off-by: Akira Moroo <retrage01@gmail.com>

retrage committed:
remove merged patches

Signed-off-by: Akira Moroo <retrage01@gmail.com>

retrage committed:
travis: add g++ to dependencies

Signed-off-by: Akira Moroo <retrage01@gmail.com>
